### ISSUE TITLE
[clang-format] wrong formatting of CUDA kernel launches (<<<...>>>) when SpacesInAngles is set to Always

### ISSUE THREAD

**Issue by lahwaacz** [llvm/llvm-project#52881]
Minimal reproducible example:
```cuda
__global__ void foo() {}

int main()
{
  foo<<<1, 1>>>();
}
```

Let's format it with clang-format (we want to use `SpacesInAngles: Always` in our project):
```
$ clang-format -style "{BasedOnStyle: mozilla, SpacesInAngles: Always}" foo.cu
__global__ void
foo()
{}

int
main()
{
  foo< < < 1, 1 > > >();
}
```
Notice that the CUDA kernel launch was re-formatted into `foo< < < 1, 1 > > >()`. OK, let's try to disable clang-format in this block:
```cuda
__global__ void foo() {}

int main()
{
  // clang-format off
  foo<<<1, 1>>>();
  // clang-format on
}
```
But then:
```
$ clang-format -style "{BasedOnStyle: mozilla, SpacesInAngles: Always}" bar.cu
__global__ void
foo()
{}

int
main()
{
  // clang-format off
  foo< <<1, 1> >>();
  // clang-format on
}
```
We got something else, but clang-format still inserted 2 spaces in the block where it should have been disabled.

Tested with clang-format version 13.0.0

**Comment by mkurdej** [comment]
It should be noted that clang-format doesn't have any special support of CUDA syntax extensions.
I consider hence this report as a request to add such support.
Apart from `<<<` kernels `>>>`, are there other CUDA-specific things to be aware of?

**Comment by lahwaacz** [comment]
I think all other CUDA-specific things are handled correctly by clang-format, or at least in a non-breaking way. Most notably, the formatting of memory and execution space specifiers (e.g. `__global__`, `__host__`, `__device__`, etc.) may depend on the style, but this can be hacked with various `*Macros` options in clang-format.

See https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#c-language-extensions for reference.

**Comment by mydeveloperday** [comment]
This is a potential fix for the clang-format issue https://reviews.llvm.org/D116494

**Comment by lahwaacz** [comment]
@mydeveloperday There are 2 additional optional arguments, i.e. there may be 2-4 arguments between `<<<` and `>>>`. I'm not sure if `{identifer|numeric_constant}` is precise, I think it's possible (though uncommon) to use e.g. expressions there as well...

Of course, there may be arbitrary number of arguments inside the parentheses (like a regular function call).

And the CUDA kernel may also be a template that can be explicitly instantiated (like a regular function template). So something like this is valid too:
```cpp
foo<bar...><<<1, 2, 3, 0>>>(args...);
```

**Comment by dmsteck** [comment]
Still a bug in clang-format 15.0.6. Repro here: https://github.com/dmsteck/clang-format-cuda-bug

It's worth noting that `<<<` and `>>>` are special tokens, much like the streaming operators `<<` and `>>` in normal C++ code. Unless I'm missing something, `<<<` (with any whitespace in between) cannot even occur in a legal C++ program. 

Given that clang-format automatically treats `.cu` files as C++ files and offers no way of specializing behaviour for `.cu` files, a fix would be really appreciated.

Here are some potential test cases...
```cpp
foo<<<1,2,3>>>();                 ->    foo<<< 1, 2, 3 >>>();
bar<int><<<1,2,3>>>();            ->    bar< int ><<< 1, 2, 3 >>>();
// If you want to go hardcore
foo<<<1, 1>>>(), 1 >> 1;         ->    foo<<< 1, 1 >>>(), 1 >> 1;
bar<int><<<1, 1>>>(), 1 >> 1;         ->    bar< int ><<< 1, 1 >>>(), 1 >> 1;
```
